Source/Core/Core/HW/BBA-TAP/TAP_Win32.cpp

```diff
@@ -241,56 +259,171 @@ bool CEXIETHERNET::IsActivated()
+		ERROR_LOG(SP1, "ReadFile failed (err=0x%X)", err);
+		return false;
 	}
 
-	// Always report the packet as being sent successfully, even though it might be a lie
-	SendComplete();
-
 	return true;
 }
 
-VOID CALLBACK CEXIETHERNET::ReadWaitCallback(PVOID lpParameter, BOOLEAN TimerFired)
+static void ReadThreadHandler(CEXIETHERNET* self)
 {
-	CEXIETHERNET* self = (CEXIETHERNET*)lpParameter;
+	// Queue an asynchronous read into the buffer in the CEXIETHERNET class.
+	// Unlike writes, there is a persistent buffer for reads, since we can only have one in-flight at once. This request
+	// is guaranteed to finish before the thread exists, so it is safe to allocate on the stack. The alternative here
+	// would be allocating a buffer for every packet that comes in, which is redundant and un-necessary.
+	AsyncIOHeader read_io_hdr;
+	bool read_io_active = QueueRead(self, &read_io_hdr);
+
+	// Run until we're deactivated.
+	while (self->mHCompletionPort != INVALID_HANDLE_VALUE)
+	{
+		DWORD bytes_transferred;
+		ULONG_PTR completion_key;
+		OVERLAPPED* overlapped;
+		BOOL result = GetQueuedCompletionStatus(self->mHCompletionPort, &bytes_transferred, &completion_key, &overlapped, INFINITE);
+
+		// GetQueuedCompletionStatus returns false if the IO request encountered an error, or was canceled.
+		// If overlapped is set to null, no packet was dequeued. See MSDN docs.
+		if (!result && !overlapped)
+		{
+			WARN_LOG(SP1, "GetQueuedCompletionStatus: err 0x%X", GetLastError());
+			continue;
+		}
+
+		// For read/write events on the file handle, completion_key is set to self.
+		// If the completion key was null, this was a "close" packet queued by CEXIETHERNET::Deactivate.
+		if (completion_key == reinterpret_cast<ULONG_PTR>(self))
+		{
+			// Calculate the address of the containing structure, ASyncIORequest, from the address of the overlapped member.
+			AsyncIOHeader* io_hdr = CONTAINING_RECORD(overlapped, AsyncIOHeader, overlapped);
+			if (io_hdr->type == AsyncIOHeader::IO_PACKET_TYPE_READ)
+			{
+				// Pass the size back to the hardware interface, and fire the interrupt.
+				_dbg_assert_(SP1, read_io_active);
+				if (self->readEnabled.load())
+				{
+					DEBUG_LOG(SP1, "Received %u bytes\n: %s", bytes_transferred, ArrayToString(self->mRecvBuffer, bytes_transferred, 0x10).c_str());
+					self->mRecvBufferLength = bytes_transferred;
+					self->RecvHandlePacket();
+				}
 
-	GetOverlappedResult(self->mHAdapter, &self->mReadOverlapped,
-		(LPDWORD)&self->mRecvBufferLength, false);
+				// Re-use the same read buffer, since it's been copied into HW memory, and queue another read request.
+				read_io_active = QueueRead(self, &read_io_hdr);
+			}
+			else if (io_hdr->type == AsyncIOHeader::IO_PACKET_TYPE_WRITE)
+			{
+				// Cast to the write buffer class, to free (if any) buffer allocated with the write.
+				// See the note in CEXIETHERNET::SendFrame for why.
+				delete static_cast<AsyncIOWritePacket*>(io_hdr);
```

Comment:
Is there no particular way to avoid needing explicit deletion here (or at least managing it in such a way a smart pointer reset() can be used instead)?